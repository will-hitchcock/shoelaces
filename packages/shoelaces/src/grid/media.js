import breakpoints, { keys } from '../const/breakpoints'
import deepmerge from 'deepmerge'

const queries = {
  min: key => `@media (min-width: ${breakpoints[key]})`,
  max: key => `@media (max-width: ${breakpoints[key]})`,
  between: (start, end) =>
    `@media (min-width: ${breakpoints[start]} and (max-width: ${breakpoints[
      end
    ]}`
}

export const min = key => args => {
  if (!Object.keys(args).length) return null
  if (!key || !breakpoints[key]) {
    return args
  }
  return { [queries.min(key)]: args }
}

export const max = key => args => {
  if (!Object.keys(args).length) return null
  if (!key) {
    return args
  }
  return { [queries.max(key)]: args }
}

export const between = (start, end) => args => {
  if (!Object.keys(args).length) return null
  if (!start || !end) {
    return args
  }
  return { [queries.between(start, end)]: args }
}

export const doQuery = (query, obj, func) => {
  let processedObj = {}

  if (typeof obj !== 'object') {
    return { ...processedObj, ...func(obj) }
  }

  for (let key in obj) {
    processedObj = { ...processedObj, ...query(key)(func(obj[key])) }
  }

  return processedObj
}

export const deepQuery = (query, arr) =>
  deepmerge.all(arr.map(item => doQuery(query, item.data, item.mixin)))

export default { min, max, between, doQuery, deepQuery }
