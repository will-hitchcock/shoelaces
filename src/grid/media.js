import breakpoints from '../const/breakpoints'

const queries = {
  min: key => `@media (min-width: ${breakpoints[key]})`,
  max: key => `@media (max-width: ${breakpoints[key]})`,
  between: (start, end) =>
    `@media (min-width: ${
      breakpoints[start]
    } and (max-width: ${
      breakpoints[end]
    }`
}

export const min = key => args => {
  if (!key || !breakpoints[key]) { return args }
  return { [queries.min(key)]: args }
}

export const max = key => args => {
  if (!key) { return args }
  return { [queries.max(key)]: args }
}

export const between = (start, end) => args => {
  if (!start || !end) { return args }
  return { [queries.between(start, end)]: args }
}

export const flatten = (query, prop, mixin) => {
  if (typeof prop !== 'object') { return mixin(prop) }
  const output = {}

  Object.keys(breakpoints)
    .filter(a => !!prop[a])
    .map(value => (output[query(value)] = mixin(prop[value])))

  return output
}

export const flattenMin = (...args) => flatten(queries.min, ...args)
export const flattenMax = (...args) => flatten(queries.max, ...args)

export default { min, max, between, flattenMin, flattenMax }
