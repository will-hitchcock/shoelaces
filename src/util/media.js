import breakpoints from '../const/breakpoints'

export const queries = {
  min: key => `@media (min-width: ${breakpoints[key]})`,
  max: key => `@media (max-width: ${breakpoints[key]})`,
  between: (start, end) =>
    `@media (min-width: ${
      breakpoints[start]
    } and (max-width: ${
      breakpoints[end]
    }`
}

export const min = key => args => {
  if (!key) return args
  return { [queries.min(key)]: args }
}

export const max = key => args => {
  if (!key) return args
  return { [queries.max(key)]: args }
}

export const between = (start, end) => args => ({
  [queries.between(start, end)]: args
})

export const flatten = (query, prop, mixin) => {
  if (typeof prop !== 'object') { return mixin(prop) }
  const output = {}

  Object.keys(breakpoints)
    .filter(a => !!prop[a])
    .map(value => (output[query(value)] = mixin(prop[value])))

  return output
}

export const flattenMin = (...args) => flatten(queries.min, ...args)
export const flattenMax = (...args) => flatten(queries.max, ...args)

export default { queries, min, max, between, flatten, flattenMin, flattenMax }
