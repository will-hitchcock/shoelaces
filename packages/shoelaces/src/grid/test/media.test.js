import media from '../media';

describe('media', () => {
  it('returns the result of the predicate function at the correct breakpoints', () => {
    expect(media((width) => ({ width }), { xs: '100%', md: '50%' })).toMatchSnapshot();
  });

  it('returns the result of the function literally if the input is not an object', () => {
    expect(media((width) => ({ width }), '100%')).toMatchSnapshot();
  });
});
