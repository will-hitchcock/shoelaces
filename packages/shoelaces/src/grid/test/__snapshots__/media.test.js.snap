// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`media returns the result of the function literally if the input is not an object 1`] = `
Object {
  "width": "100%",
}
`;

exports[`media returns the result of the predicate function at the correct breakpoints 1`] = `
Object {
  "@media (min-width: 0)": Object {
    "width": "100%",
  },
  "@media (min-width: 48em)": Object {
    "width": "50%",
  },
}
`;
